File: browser/lib/session.ts
import { ISessionFactory, ISession } from "./types";

const tables = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXTZabcdefghiklmnopqrstuvwxyz".split("");

export class SessionFactory implements ISessionFactory {
    private readonly __session: Record<string, ISession> = {};
    constructor(private readonly __id: string) {

    }

    public id(): string {
        return this.__id;
    }

    public attach(id: string,conn: RTCPeerConnection): ISession{
        (conn as any).state = function (): Promise<number>{
            return new Promise(function (resolve,reject) {
                this.__resolve = resolve;
            }.bind(this));
        };
        (conn as any).id = function () {
            return id;
        };
        const __close = conn.close.bind(conn);
        (conn as any).close = () => {
            this.remove(id);
            __close();
        };
        this.__session[id] = conn as ISession;
        return conn as ISession;
    }

    public create(id: string,conn: RTCPeerConnection): ISession {
        return this.attach(this.__random(id),conn);
    }

    public query(id: string): ISession {
        return this.__session[id];
    }

    public remove(id: string): void {
        delete (this.__session[id]);
    }

    public __random(id: string, size: number = 8): string {
        if (!length) {
            length = Math.floor(Math.random() * tables.length);
        }
        let str = "";
        for (let i = 0; i < size; i++) {
            str += tables[Math.floor(Math.random() * tables.length)];
        }
        return [this.__id, id, str].join(".");
    }
}

File: peer/session.go
package peer

import (
	"fmt"
	"github.com/pions/webrtc"
	"math/rand"
	"time"
)

func NewSessionFactory(id string) SessionFactory {
	return &CoreSessionFactory{
		id:     id,
		source: rand.NewSource(time.Now().Unix()),
	}
}

func (f *CoreSessionFactory) Id() string {
	return f.id
}

func (f *CoreSessionFactory) __id(id string) string {
	return fmt.Sprintf("%s.%s.%d", f.id, id, f.source.Int63())
}

func (f *CoreSessionFactory) Query(id string) *Session {
	it, ok := f.table.Load(id)
	if ok {
		return it.(*Session)
	}
	return nil
}

func (f *CoreSessionFactory) Remove(id string) {
	f.table.Delete(id)
}

func (f *CoreSessionFactory) Attach(id string, conn *webrtc.RTCPeerConnection) *Session {
	session := NewSession(id, conn)
	f.table.Store(id, session)
	return session
}

func (f *CoreSessionFactory) Create(id string, conn *webrtc.RTCPeerConnection) *Session {
	return f.Attach(f.__id(id), conn)
}

func (s *Session) Close() error {
	err := s.RTCPeerConnection.Close()
	close(s.done)
	return err
}

func NewSession(id string, conn *webrtc.RTCPeerConnection) *Session {
	return &Session{
		RTCPeerConnection: conn,
		id:                id,
		state:             make(chan int, 1),
		done:              make(chan error, 1),
	}
}
func (s *Session) Id() string {
	return s.id
}

func (s *Session) Done() error {
	return <-s.done
}

func (s *Session) State() int {
	return <-s.state
}
